### ISSUE TITLE
Carbon-cache keeps accepting connections after metrics are paused; hits file descriptor limit and fills up the disk

### ISSUE THREAD

**Issue by cataphract** [graphite-project/carbon#428]
If, for whatever reason, the MetricCache gets full, carbon-cache will pause the metrics (at least I can see in the debugger that `state.metricReceiversPaused = true` and the only event I saw in the code that triggered that state change was "cache full", but I never observed the transition, so the trigger may be elsewhere). This will cause the newly accepted connections to be removed from the `epoll`, hence any events from those connections will not be received.

Carbon-cache should stop `accept()`ing connections. Because it does otherwise, and removes the accepted file descriptors from poll (making it unaware of disconnects by the client that would otherwise trigger the file descriptor to be closed) the number of open file descriptors will keep increasing until the limit is hit. At this point, most of these connections will be in a `CLOSE_WAIT` state because the client will already have closed the connection on its end.

But it gets worse. Once the file descriptor limit is hit, carbon-cache will write to the log at a very high rate:

<pre>10/09/2015 10:35:14 :: Could not accept new connection (EMFILE)</pre>

which will end up filling up the disk.

**Comment by jkielbaey** [comment]
I ran into a similar issue. During a planned intervention, all TCP connection states were dropped on a firewall. As a result about 700 diamond agents lost their connection to `carbon-relay` and tried to open a new connection. Since the old connections were not properly closed, carbon-relay still had 700 old/stale connections.

The new connections pushed carbon-relay over the limit of its max open file descriptors (1024).
As a result carbon-relay fills up the logs (53 messages / sec) and the carbon-relay process consumes 100% CPU.
It would indeed be better if carbon-relay doesn't accept any new connections once it reaches this limit and also throttle the logging of this message.

**Comment by mleinart** [comment]
I'm not sure what we can really do here. The log message is fired by Twisted, not Graphite: https://github.com/twisted/twisted/blob/trunk/twisted/internet/tcp.py#L1045-L1063

As far as not accepting new connections while the transport is paused, this is probably best done through a feature request to Twisted. We could call transport.loseConnection(), but that would (I think) also kill existing but paused TCP connections.

*PERHAPS* we could set transport.disconnecting = True to keep accept()s from happening (https://github.com/twisted/twisted/blob/trunk/twisted/internet/tcp.py#L1032-L1035) but that's certainly unsupported and prone to breakage.


Guessing our best way forward is to
a) Document somewhere a recommendation to set a high nofiles ulimit (1024 is quite low for any high volume TCP server)
b) Log the current ulimit at startup and/or start recording current open FDs in carbon metrics to bring it front-of-mind to people administering carbon.

**Comment by NickSchaf** [comment]
I took a quick glance through code yesterday and also noticed this seemed to be in Twisted.  However, I'll include my use-case and comments just for more data on the issue (and in case something else could be done in Graphite to compensate):

A high ulimit just delays the problem - this is still equivalent to a memory leak.  I have a test network with 16 devices logging metrics to Graphite.  It's not burdensome: each device reports its set of metrics (1.5-3KB message size per device), every minute.  The problem is the network is wireless and very lossy - some links are very weak.  I'm using TCP to improve reliability, but connections are still lost occasionally.  It took a few weeks to run out of 1024 FDs.  Very much like what @jkielbaey mentioned - all those old, dead connections were consuming FDs.

Also mentioned by @jkielbaey were the logs.  I was out of file descriptors for about a day before noticing the problem; the logs for that day consumed almost 200GB.

As a work-around, I'll be implementing a monitor to restart carbon-aggregator when its established TCP connections count grows too high.  I plan on increasing the size of my test network, so the problem will just surface more frequently without such a monitor.

I do like b) presented by @mleinart - I actually looked through the auto-created metrics yesterday to see whether that happened to be there already.
